power_grid_model_c/power_grid_model_c/include/power_grid_model_c/dataset.h

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * @brief destroy an instance of PGM_ConstDataset
+ * @param dataset a pointer to the PGM_ConstDataset created by PGM_create_const_dataset()
+ * @return
+ */
+PGM_API void PGM_destroy_const_dataset(PGM_ConstDataset* dataset);
+
+/**
+ * @brief Add a component buffer to an instance of PGM_ConstDataset
+ * @param handle
+ * @param dataset a pointer to the PGM_ConstDataset
+ * @param component name of the component
+ * @param elements_per_scenario numbers of the elements per scenario
+ *     If the component is uniform, elements_per_scenario must be >= 0
+ *     If the component is not uniform, elements_per_scenario must be -1
+ * @param total_elements total number of elements for all scenarios
+ *     If elements_per_scenario >= 0, we must have elements_per_scenario * batch_size = total_elements
+ * @param indptr pointer to array of indptr of a non-uniform component
+ *     If the component is uniform, indptr must be NULL.
+ *     If the component is not uniform, indptr must point to an array of size (batch_size + 1)
+ *         The values in the array must be not decreasing.
+ *         And we must have indptr[0] = 0, indptr[batch_size] = total_elements
+ * @param data void pointer to the buffer data
+ * @return
+ */
+PGM_API void PGM_const_dataset_add_buffer(PGM_Handle* handle, PGM_ConstDataset* dataset, char const* component,
+                                          PGM_Idx elements_per_scenario, PGM_Idx total_elements, PGM_Idx const* indptr,
+                                          void const* data);
+
+/**
+ * @brief Get the dataset info of the instance PGM_ConstDataset
+ * @param handle
+ * @param dataset a pointer to the PGM_ConstDataset
+ * @return pointer to the instance of PGM_DatasetInfo.
+ *     The pointer has the same lifetime as the input dataset pointer.
+ */
+PGM_API PGM_DatasetInfo const* PGM_const_dataset_get_info(PGM_Handle* handle, PGM_ConstDataset const* dataset);
+
+/**
+ * @brief Get the dataset info of the instance PGM_WritableDataset
+ * @param handle
+ * @param dataset a pointer to the PGM_WritableDataset
+ * @return pointer to the instance of PGM_DatasetInfo.
+ *     The pointer has the same lifetime as the input dataset pointer.
+ */
+PGM_API PGM_DatasetInfo const* PGM_writable_dataset_get_info(PGM_Handle* handle, PGM_WritableDataset const* dataset);
+
+/**
+ * @brief Set buffer into the instance PGM_WritableDataset
+ * @param handle
+ * @param dataset a pointer to the PGM_WritableDataset
+ * @param component name of the component
+ * @param indptr pointer to array of indptr of a non-uniform component
+ *     If the component is uniform, indptr must be NULL.
+ *     If the component is not uniform, indptr must point to an array of size (batch_size + 1)
+ * @param data void pointer to the buffer data
+ * @return
+ */
+PGM_API void PGM_writable_dataset_set_buffer(PGM_Handle* handle, PGM_WritableDataset* dataset, char const* component,
+                                             PGM_Idx* indptr, void* data);
```

Comment:
how does one create a `PGM_WritableDataset`? there's no `create` function for it

Reply:
You are not supposed to create it by yourself. The deserializer will create one when it deserialize the json/msgpack data. You will get the dataset, and set buffer into it.